src/Compilers/CSharp/Portable/Symbols/TypeMap.cs

```diff
@@ -42,7 +42,7 @@ internal TypeMap(ImmutableArray<TypeParameterSymbol> from, ImmutableArray<TypeWi
             : base(ConstructMapping(from, to))
         {
             // mapping contents are read-only hereafter
-            Debug.Assert(allowAlpha || !from.Any(static tp => tp is SubstitutedTypeParameterSymbol));
+            Debug.Assert(allowAlpha || !from.Any(static tp => tp is SubstitutedTypeParameterSymbol && tp.ContainingSymbol is not SourceExtensionImplementationMethodSymbol));
```

Comment:
>tp.ContainingSymbol is not SourceExtensionImplementationMethodSymbol [](http://example.com/codeflow?start=102&length=68)

I think it would be better to check if the type parameter is original definition. Those should be fine to substitute regardless what the container is and the check would be more robust long term.